### PR TITLE
Configuration changes to the Ansible/Hadoop deploy for Dynamic Allocation & Multi user 

### PR DESCRIPTION
**Pull request by stvoutsin**
This PR includes changes to the Spark & Hadoop configuration files to enable the Dynamic Allocation feature of Yarn & Spark and to allow multiple users to run jobs concurrently.

Also addresses issue #392 , by creating an interpreters.json configuration file that sets the Spark isolation level to "isolated" "per User".
This way each user gets their own Spark Context (& their own Yarn Application), and this is isolated from other users.

With the combination of this change and the Dynamic Allocation changes, each new user is able to have their job start immediately without being stuck in a queue, and Yarn & Spark manage and shuffle the tasks around dynamically, to make this possible.

Along with issue #392 this PR also narrows down the list of interpreters that a user sees, to only the specific ones we support:
sh, python, spark &  md

### CODE REVIEW

File: deployments/common/zeppelin/interpreter.json [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+          "type": "number"
+        },
+        "zeppelin.spark.importImplicit": {
+          "name": "zeppelin.spark.importImplicit",
+          "value": true,
+          "type": "checkbox"
+        },
+        "zeppelin.dep.additionalRemoteRepository": {
+          "name": "zeppelin.dep.additionalRemoteRepository",
+          "value": "spark-packages,http://dl.bintray.com/spark-packages/maven,false;",
```
Comment (Zarquan) [inline deployments/common/zeppelin/interpreter.json:381]: Two things to check:

1. It looks like this service is being replaced
   * https://bintray.com/ _"Thanks for supporting Bintray! This service has now been sunset, and to assist with migration to the JFrog Platform, paid accounts can login until July 4th."_

2. This URL [https://bintray.com/pwendell/spark-packages] points to a test account set up by someone using a Gmail address. No additional details about who, why or what this contains.

Does having this property set in our configuration mean that Spark will trust any Java packages downloaded from this location ?

File: deployments/common/zeppelin/interpreter.json [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+        "updatePolicy": "daily",
+        "checksumPolicy": "warn"
+      },
+      "mirroredRepositories": [],
+      "repositoryManager": false
+    },
+    {
+      "id": "local",
+      "type": "default",
+      "url": "file:///home/fedora/.m2/repository",
```
Comment (Zarquan) [inline deployments/common/zeppelin/interpreter.json:616]: This hard-codes the name of the `fedora` user account into the Zeppelin configuration file.

* When we assign a separate user account for running Zeppelin this would break.
* The K8s deployment does not have a `fedora` user account.

Might be better to set this to a fixed location outside the user home directory tree, e.g. `/var/local/zeppelin/maven` which is created as part of the Ansible deployment.

File: deployments/hadoop-yarn/ansible/20-install-spark.yml [hunk 3]
```diff
@@ -33,7 +33,7 @@
     - name: "Download and unpack the {{spname}} tar gzip file"
       become: true
       unarchive:
-        src:  "https://downloads.apache.org/spark/{{spname}}/{{spfull}}.tgz"
+        src:  "https://archive.apache.org/dist/spark/{{spname}}/{{spfull}}.tgz"
```
Comment (Zarquan) [inline deployments/hadoop-yarn/ansible/20-install-spark.yml:36]: Hasn't this already been done in another PR ?

File: deployments/hadoop-yarn/ansible/29-install-pip-libs.yml [hunk 4]
```diff
@@ -27,9 +27,19 @@
...
+    - name: Copy pip requirements file into tmp
+      become: true
+      copy:
+        src: "{{ playbook_dir | dirname | dirname }}/common/pip/requirements.txt"
+        dest: "/tmp/requirements.txt"
+
+
+    - name: Install the required Python packages
       become: true
-      command: /usr/bin/pip3 install numpy scipy matplotlib grpcio jupyter cython protobuf pandas healpy astropy scikit-learn hdbscan pyvo astroquery
```
Comment (Zarquan) [inline deployments/hadoop-yarn/ansible/29-install-pip-libs.yml:34]: hasn't this already been done in PR #463 ?

File: deployments/hadoop-yarn/ansible/34-setup-shuffler.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#     GNU General Public License for more details.
+#
+#     You should have received a copy of the GNU General Public License
+#     along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#   </meta:licence>
+# </meta:header>
+#
+#
+
+- name: "Copy Shuffle jar to tmp directory"
```
Comment (Zarquan) [inline deployments/hadoop-yarn/ansible/34-setup-shuffler.yml:23]:
```suggestion
- name: "Fetch the spark-yarn-shuffle jar from one of the master nodes and store it in our /tmp directory"
```

File: deployments/hadoop-yarn/ansible/34-setup-shuffler.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  hosts: masters
```
Comment (Zarquan) [inline deployments/hadoop-yarn/ansible/34-setup-shuffler.yml:24]:
```suggestion
  hosts: master01
```

File: deployments/hadoop-yarn/ansible/34-setup-shuffler.yml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  tasks:
+    - fetch:
+        src: /opt/spark/yarn/spark-2.4.7-yarn-shuffle.jar
+        dest: /tmp/yarnspark-2.4.7-yarn-shuffle.jar
```
Comment (Zarquan) [inline deployments/hadoop-yarn/ansible/34-setup-shuffler.yml:28]: This hard codes the Spark version into this Ansible script.

If the [spname](https://github.com/wfau/aglais/blob/10a106c6493cf208e7c01899683c2b20e301297a/deployments/hadoop-yarn/ansible/config/cclake-medium-04.yml#L43) setting for this deployment changes, then  this Ansible script will look for the wrong jar file.

By design different deployments can have different values for [spname](https://github.com/wfau/aglais/blob/10a106c6493cf208e7c01899683c2b20e301297a/deployments/hadoop-yarn/ansible/config/cclake-medium-04.yml#L43).

File: notes/stv/20210617-multi-user-01.txt [hunk 8]
```diff
@@ -0,0 +1,573 @@
...
+
+# Repeat step from above, running the notebooks for each user concurrently
+
+# Both notebooks start successfully..
+# After checking yarn UI, we see 47.1% usage for each application (two applications started, 1 per user)
+
+# Let's try killing one of the two applications, to see what happens
+# Restart Spark interpreter for user #2
+
+# Application for user #1 now jumps to 95% usage, so it looks like the load balancing is working
```
Comment (Zarquan) [inline notes/stv/20210617-multi-user-01.txt:572]: Yay - success :-)
This is really useful to learn about all this.

File: notes/stv/20210617-multi-user-01.txt [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+	FINISHED   
+	Took 5 min 31 sec. Last updated by admin at June 17 2021, 11:56:43 AM.
+
+        [Success]
+
+
+# This works, so in theory we could set the maximum resource allocation per user to  set amount, as a fraction of the total resources.
+
+# However perhaps this is not the most efficient way to have a multi user setup
+# Unless we expect multiple concurrent users running jobs at the same time, our resources will be underutilized
```
Comment (Zarquan) [inline notes/stv/20210617-multi-user-01.txt:422]: Useful to learn about this.

File: notes/stv/20210628-multi-user-deploy-01.txt [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+     /AglaisPublicExamples/Good astrometric solutions via ML Random Forrest classifier   
+        https://raw.githubusercontent.com/wfau/aglais-notebooks/main/2G5NU6HTK/note.json
+	
+        [Success]
+  
+
+
+# While observing the resource usage using the Spark UI, we see two applications running, each with 47.1% usage
+
+# Also worth noting that both notebooks took about the same time to complete
```
Comment (Zarquan) [inline notes/stv/20210628-multi-user-deploy-01.txt:153]: :+1: good notes explain what is happening and how we arrived at this result.

**Review by Zarquan — CHANGES_REQUESTED** [review]
Need to resolve overlap with PR #463 and fix the hard coded paths and version numbers.

**Comment by stvoutsin** [comment]
I've made a couple of additional commits to address the points raised. 
See: 

- https://github.com/wfau/aglais/pull/517/commits/d1e088890b527015ae7af57747b0329c7f3479bb
- https://github.com/wfau/aglais/pull/517/commits/53f7ca45cbfb1f9efc332dfaa2185fed64574915
- https://github.com/wfau/aglais/pull/517/commits/90882b86ffd571c14d6e8d05f4bf8a6e7b38d1f4
- https://github.com/wfau/aglais/pull/517/commits/b0d90933b93e80f4ea2f1d12da2d1061a43731ea

I've also run a validation test and the test notebooks passed (Including multi user concurrent notebook run)

File: deployments/common/zeppelin/interpreter.json (hunk 2 above)
Comment (Zarquan) [reply deployments/common/zeppelin/interpreter.json:616]:
```suggestion
      "url": "file:///var/local/zeppelin/maven",
```
or
```suggestion
      "url": "{{zepmavendest}}",
```

File: deployments/hadoop-yarn/ansible/35-config-zeppelin-interpreters.yml [hunk 11]
```diff
@@ -20,20 +20,24 @@
...
+        src: "{{ playbook_dir | dirname | dirname }}/common/zeppelin/interpreter.json"
+        dest: "{{zephome}}/conf/interpreter.json"
 
 
+    - replace:
+        path: "{{zephome}}/conf/interpreter.json"
+        regexp: "^file:///home/fedora/.m2/repository$"
+        replace: "{{ zepmavendest }}"
+      delegate_to: zeppelin
+
```
Comment (Zarquan) [inline deployments/hadoop-yarn/ansible/35-config-zeppelin-interpreters.yml:43]: This feels really fragile.

* If someone reads the config file in deployments/common they would assume the maven directory is in the fedora user's home directory.
* If someone makes a tiny change to the value in the config file in deployments/common it won't match the regex expression any more.

We own the config file this comes from, `deployments/common/zeppelin/interpreter.json` , so why not edit the original config and treat it as a template?

File: notes/stv/20210628-multi-user-deploy-01.txt (hunk 10 above)
Comment (Zarquan) [reply notes/stv/20210628-multi-user-deploy-01.txt:153]: It would be good to have a single user run using the same configuration to act as a comparison.

* user A only - Took 5 min 50 sec
* user B only - Took 5 min 50 sec
* user A and B concurrently - Took 5 min 50 sec

Comment (Zarquan) [reply notes/stv/20210628-multi-user-deploy-01.txt:153]: Can we run the same set of tests from the command line using curl ?
That would make the tests much more reproducible.

File: notes/stv/20210628-multi-user-deploy-01.txt [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+	> Done
+
+
+
+# ----------------------------------------------------------------
+# Login via Firefox, in two separate windows, as two separate users
+#[user@desktop]
+
+    firefox --new-window "http://128.232.227.136:8080/" &
+
```
Comment (Zarquan) [inline notes/stv/20210628-multi-user-deploy-01.txt:90]: Does this work ?
Not sure you can have two windows in the same browser logged in as different users.
When I tried it the second login changed the session cookie invalidating the first login.

I think we would either need to use a different browser for each user or use something like [container tabs](https://support.mozilla.org/en-US/kb/containers) in Firefox to manage the session cookies for each user.

**Review by Zarquan — CHANGES_REQUESTED** [review]

File: deployments/hadoop-yarn/ansible/20-install-spark.yml (hunk 3 above)
Comment (Zarquan) [reply deployments/hadoop-yarn/ansible/20-install-spark.yml]: Merging master back into the fork seems to have fixed it this time, but I think [`git rebase`](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase) is the better way to fix something like this.

File: deployments/common/pip/requirements.txt [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+jupyter==1.0.0
+Cython==0.29.23
+protobuf==3.16.0
+pandas==1.2.4
+healpy==1.14.0
+astropy==4.2.1
+astroquery==0.4.1
+scikit-learn==0.24.2
+hdbscan==0.8.27
+pyvo==1.1
```
[the remark this answers is not in the document]
Comment (Zarquan) [reply deployments/common/pip/requirements.txt]: Can we confirm that this is no longer part of the pull request ?
If merging master back in to this branch resolved it, then why is this still showing up in the  pull request as a change for review ?

File: deployments/hadoop-yarn/ansible/20-install-spark.yml (hunk 3 above)
Comment (Zarquan) [reply deployments/hadoop-yarn/ansible/20-install-spark.yml]: Can we confirm that this is no longer part of the pull request ?
If merging master back in to this branch resolved it, then why is this still showing up in the  pull request as a change for review ?

File: deployments/hadoop-yarn/ansible/29-install-pip-libs.yml (hunk 4 above)
Comment (Zarquan) [reply deployments/hadoop-yarn/ansible/29-install-pip-libs.yml]: Can we confirm that this is no longer part of the pull request ?
If merging master back in to this branch resolved it, then why is this still showing up in the  pull request as a change for review ?

File: notes/stv/20210628-multi-user-deploy-01.txt (hunk 12 above)
Comment (stvoutsin) [reply notes/stv/20210628-multi-user-deploy-01.txt:90]: This works if you start it as an incognito window

File: deployments/common/pip/requirements.txt (hunk 13 above)
[the remark this answers is not in the document]
Comment (stvoutsin) [reply deployments/common/pip/requirements.txt]: Looking at the files changed page https://github.com/wfau/aglais/pull/517/files it doesn't look like this file is part of the files changed, so it looks ok to me. Otherwise I'm not sure what to do to have it not appear in the changes.

File: deployments/hadoop-yarn/ansible/20-install-spark.yml (hunk 3 above)
Comment (stvoutsin) [reply deployments/hadoop-yarn/ansible/20-install-spark.yml]: Same as above, I don't see this in the changed files of the PR https://github.com/wfau/aglais/pull/517/files

File: deployments/common/pip/requirements.txt (hunk 13 above)
[the remark this answers is not in the document]
Comment (Zarquan) [reply deployments/common/pip/requirements.txt]: OK, thanks for checking.

File: deployments/hadoop-yarn/ansible/20-install-spark.yml (hunk 3 above)
Comment (Zarquan) [reply deployments/hadoop-yarn/ansible/20-install-spark.yml]: OK, thanks for checking.

File: notes/stv/20210628-multi-user-deploy-01.txt (hunk 10 above)
Comment (Zarquan) [reply notes/stv/20210628-multi-user-deploy-01.txt:153]: This is not a shop stopper - we can merge without this.

**Comment by stvoutsin** [comment]
I've added some more changes based on the suggestions, I will run some tests tomorrow and post the results as another commit

File: notes/stv/20210628-multi-user-deploy-01.txt (hunk 10 above)
Comment (Zarquan) [reply notes/stv/20210628-multi-user-deploy-01.txt:154]: OK - not a show stopper, but without this comparison your statement that _"both notebooks took about the same time to complete"_ doesn't have enough evidence.

File: notes/stv/20210628-multi-user-deploy-01.txt [hunk 14]
```diff
@@ -88,7 +88,8 @@
 
     firefox --new-window "http://128.232.227.136:8080/" &
 
-
+    # Open second window as Incognito user, so that we can login as a different user
+
```
Comment (Zarquan) [inline notes/stv/20210628-multi-user-deploy-01.txt:92]: Adding the commandline options for this would make it easier to reproduce it.

**Review by Zarquan — APPROVED** [review]
Yep - ship it :+1:
